libs/langchain-community/src/chat_models/perplexity.ts

```diff
@@ -0,0 +1,290 @@
+
+  frequencyPenalty = 1;
+
+  private client: OpenAI;
+
+  constructor(fields?: Partial<PerplexityChatInput> & BaseChatModelParams) {
+    super(fields ?? {});
+
+    this.modelName = fields?.modelName ?? this.modelName;
+    this.temperature = fields?.temperature ?? this.temperature;
+    this.maxTokens = fields?.maxTokens;
+    this.apiKey =
+      fields?.apiKey ?? getEnvironmentVariable("PERPLEXITY_API_KEY");
+    this.streaming = fields?.streaming ?? this.streaming;
+    this.timeout = fields?.timeout;
+    this.topP = fields?.topP ?? this.topP;
+    this.returnImages = fields?.returnImages ?? this.returnImages;
+    this.returnRelatedQuestions =
+      fields?.returnRelatedQuestions ?? this.returnRelatedQuestions;
+    this.topK = fields?.topK ?? this.topK;
+    this.presencePenalty = fields?.presencePenalty ?? this.presencePenalty;
+    this.frequencyPenalty = fields?.frequencyPenalty ?? this.frequencyPenalty;
+
+    if (!this.apiKey) {
+      throw new Error("Perplexity API key not found");
+    }
+
+    this.client = new OpenAI({
+      apiKey: this.apiKey,
+      baseURL: "https://api.perplexity.ai",
+    });
+  }
+
+  _llmType() {
+    return "perplexity";
+  }
+
+  /**
+   * Get the parameters used to invoke the model
+   */
+  invocationParams() {
+    return {
+      model: this.modelName,
+      temperature: this.temperature,
+      max_tokens: this.maxTokens,
+      stream: this.streaming,
+      top_p: this.topP,
+      return_images: this.returnImages,
+      return_related_questions: this.returnRelatedQuestions,
+      top_k: this.topK,
+      presence_penalty: this.presencePenalty,
+      frequency_penalty: this.frequencyPenalty,
+    };
+  }
+
+  /**
+   * Convert a message to a format that the model expects
+   */
+  private messageToPerplexityRole(message: BaseMessage): {
+    role: PerplexityRole;
+    content: string;
+  } {
+    if (message instanceof ChatMessage) {
+      return {
+        role: message.role as PerplexityRole,
+        content: message.content.toString(),
+      };
+    }
+    if (message instanceof HumanMessage) {
+      return {
+        role: "user",
+        content: message.content.toString(),
+      };
+    }
+    if (message instanceof AIMessage) {
+      return {
+        role: "assistant",
+        content: message.content.toString(),
+      };
+    }
+    if (message instanceof SystemMessage) {
+      return {
+        role: "system",
+        content: message.content.toString(),
+      };
+    }
+    throw new Error(`Unknown message type: ${message}`);
+  }
+
+  async _generate(
+    messages: BaseMessage[],
+    _options?: this["ParsedCallOptions"],
+    _runManager?: CallbackManagerForLLMRun
+  ): Promise<ChatResult> {
+    const messagesList = messages.map((message) =>
+      this.messageToPerplexityRole(message)
+    );
+
+    const response = await this.client.chat.completions.create({
+      messages: messagesList,
+      ...this.invocationParams(),
+      stream: false,
```

Comment:
We do also support calling `.invoke`/`.generate` with `stream: true` and aggregating

This is niche but convenient if you want to use the final output while also streaming back tokens